Replace "GitHub Enterprise Server" option with "other" in gh auth login prompting

jtmcg opened the pull request:
Closes #9641

## Changes

- `gh auth login` prompting now provides an "other" option in place of "GitHub Enterprise Server"
- Tests were added/updated to cover this change
- Docs were updated for the `gh auth login` command to clarify how to use hostnames

## To test

1. Pull down the branch and build using `make`
2. Run `bin/gh auth login`
3. Notice the change in prompt

<img width="714" alt="Screenshot 2024-09-20 at 2 16 30 PM" src="https://github.com/user-attachments/assets/ea8905af-a66e-4866-b265-bc9e4a63831d">

4. Complete the flow to see that it is successful

<img width="810" alt="Screenshot 2024-09-20 at 2 20 42 PM" src="https://github.com/user-attachments/assets/d37c94a4-20a6-486c-9706-c209ab8c045f">

jtmcg committed:
Replace "GitHub Enterprise Server" with "other" in `gh auth login` prompt

This change is meant to better support the login flow for other customers
besides GitHub Enterprise Server customers that use the same login flow
as GHES.

jtmcg committed:
Add documentation explaining how to use `hostname` for `gh auth login`

jtmcg commented on pkg/cmd/auth/login/login.go:
```diff
@@ -59,6 +59,10 @@ func NewCmdLogin(f *cmdutil.Factory, runF func(*LoginOptions) error) *cobra.Comm
 		Long: heredoc.Docf(`
 			Authenticate with a GitHub host.
 
+			Your %[1]shostname%[1]s is the domain where the GitHub instance you are logging into is hosted.
+			For example, "github.com" is the default hostname for GitHub.com. To authenticate with
+			a different hostname like github.enterprise.com, use that hostname instead.
+
```
Looking for feedback, here 🙂

mxie replied:
There's some weird or awkward about this description and I wonder if it's because of the term `hostname`. Technically, `hostname` can be part of a domain. Here's what I'm thinking:

```suggestion
			The %[1]shost%[1]s is the domain where the GitHub instance you are logging into is hosted.
			For example, "github.com" is the where GitHub is hosted. To authenticate with
			a different host like github.enterprise.com, enter that domain instead.

```

mxie replied:
Seeing the rest of the PR, I realize that my suggestion would also change the scope a bit since `hostname` already existed as a prompt. 😅 In that case, maybe:

```suggestion
			The %[1]shostname%[1]s is the domain where the GitHub instance you are logging into is hosted.
			For example, "github.com" is the hostname for GitHub.com. To authenticate with
			a different hostname like github.enterprise.com, use that hostname instead.

```

BagToad replied:
This seems very verbose. I think we could keep it simpler. If we want to give examples, we should update/add examples to the examples section around L89. We already have an example for using `--hostname` but it could do with a revision maybe:

```
EXAMPLES
  # Start interactive setup
  $ gh auth login

  # Authenticate against github.com by reading the token from a file
  $ gh auth login --with-token < mytoken.txt

  # Authenticate with specific host
  $ gh auth login --hostname enterprise.internal
```

Anyways, here's my suggestion for this part. I don't LOVE this, but I'd vote for a simpler outcome here regardless if my suggestion is accepted or not 😆 

```suggestion
			Your %[1]shostname%[1]s is where you login to GitHub. The default hostname is GitHub.com.

```

jtmcg replied:
I do like simpler... Do you think any folks will get confused without the example?

BagToad replied:
Personally, I don't think anyone will be confused without the example. Folks needing to provide a hostname, such as GHES users, should already be familiar with the idea of providing a hostname that isn't GitHub.com for interacting with their appliance. 

There's also that example that already exists for `--hostname` usage lower in the doc. If anything, I think improving that example could be worthwhile.

But if it was my decision, I'd just keep it simple for now :)

jtmcg replied:
That's compelling. I've shortened it to your change, here. I think the example down below in the code clears up anything that might be unclear with this description. Thanks for the feedback!

mxie reviewed (commented):
Left some thoughts on the docs!

BagToad commented on pkg/cmd/auth/login/login.go:
```diff
@@ -222,7 +226,7 @@ func loginRun(opts *LoginOptions) error {
 
 func promptForHostname(opts *LoginOptions) (string, error) {
-	options := []string{"GitHub.com", "GitHub Enterprise Server"}
+	options := []string{"GitHub.com", "other"}
```
Should we sentence case `other`? 

I know the A/C has it uncapitalized, but it seems like there might be some precedent established to capitalize the first letter of prompt options. Example:

```
? How would you like to authenticate GitHub CLI?  [Use arrows to move, type to filter]
> Login with a web browser
  Paste an authentication token
```

[I checked Primer as well](https://primer.style/native/cli/getting-started#make-it-feel-like-github), and that seems to suggest we should use sentence case where we can:

![image](https://github.com/user-attachments/assets/a9c6cbd8-8fef-412d-a9ec-5da82844c7f0)

[This Primer example](https://primer.style/native/cli/components#radio-select) for a radio prompt also uses sentence case:

![image](https://github.com/user-attachments/assets/ad570f06-fb61-42f6-a700-312023ba2c77)

williammartin reviewed (changes requested):
I think this is missing both prompt changes from the A/C:

```
➜  gh auth login
? Where do you use GitHub? [Use arrows to move, type to filter]
  github.com
> other

? Hostname:
```

```
? What account do you want to log into?  [Use arrows to move, type to filter]
? Where do you use GitHub? [Use arrows to move, type to filter]
```

```
? GHE Hostname:
? Hostname:
```

jtmcg commented:
> I think this is missing both prompt changes from the A/C:
> 
> ```
> ➜  gh auth login
> ? Where do you use GitHub? [Use arrows to move, type to filter]
>   github.com
> > other
> 
> ? Hostname:
> ```
> 
> ```
> ? What account do you want to log into?  [Use arrows to move, type to filter]
> ? Where do you use GitHub? [Use arrows to move, type to filter]
> ```
> 
> ```
> ? GHE Hostname:
> ? Hostname:
> ```

Totally missed that detail! Thanks for calling them out.

jtmcg committed:
Sentence case 'Other' option in hostname prompt

jtmcg committed:
Change prompts for `gh auth login` to reflect change from GHE to Other

jtmcg committed:
Update language on docstring for `gh auth login`

jtmcg committed:
Shorten language on Authenticate with a GitHub host.

The default authentication mode is a web-based browser flow. After completion, an
authentication token will be stored securely in the system credential store.
If a credential store is not found or there is an issue using it gh will fallback
to writing the token to a plain text file. See `gh auth status` for its
stored location.

Alternatively, use `--with-token` to pass in a token on standard input.
The minimum required scopes for the token are: `repo`, `read:org`, and `gist`.

Alternatively, gh will use the authentication token found in environment variables.
This method is most suitable for "headless" use of gh such as in automation. See
`gh help environment` for more info.

To use gh in GitHub Actions, add `GH_TOKEN: ${{ github.token }}` to `env`.

The git protocol to use for git operations on this host can be set with `--git-protocol`,
or during the interactive prompting. Although login is for a single account on a host, setting
the git protocol will take effect for all users on the host.

Specifying `ssh` for the git protocol will detect existing SSH keys to upload,
prompting to create and upload a new key if one is not found. This can be skipped with
`--skip-ssh-key` flag.

USAGE
  gh auth login [flags]

FLAGS
  -p, --git-protocol string   The protocol to use for git operations on this host: {ssh|https}
  -h, --hostname string       The hostname of the GitHub instance to authenticate with
      --insecure-storage      Save authentication credentials in plain text instead of credential store
  -s, --scopes strings        Additional authentication scopes to request
      --skip-ssh-key          Skip generate/upload SSH key prompt
  -w, --web                   Open a browser to authenticate
      --with-token            Read token from standard input

INHERITED FLAGS
  --help   Show help for command

EXAMPLES
  # Start interactive setup
  $ gh auth login

  # Authenticate against github.com by reading the token from a file
  $ gh auth login --with-token < mytoken.txt

  # Authenticate with specific host
  $ gh auth login --hostname enterprise.internal

LEARN MORE
  Use `gh <command> <subcommand> --help` for more information about a command.
  Read the manual at https://cli.github.com/manual
  Learn about exit codes using `gh help exit-codes` around Tylers-GitHub-MacBook.local

williammartin commented:
**Given** I am in an interactive terminal environment
**When** I run gh auth login
**Then** I am presented with:

```
➜  gh auth login
? Where do you use GitHub? [Use arrows to move, type to filter]
  github.com
> other

? Hostname:
```

---

```
➜ ./bin/gh auth login
? Where do you use GitHub?  [Use arrows to move, type to filter]
  GitHub.com
> Other
```

```
➜ ./bin/gh auth login
? Where do you use GitHub? Other
? Hostname:
```

williammartin commented on pkg/cmd/auth/login/login.go:
```diff
@@ -59,6 +59,8 @@ func NewCmdLogin(f *cmdutil.Factory, runF func(*LoginOptions) error) *cobra.Comm
 		Long: heredoc.Docf(`
 			Authenticate with a GitHub host.
 
+			The %[1]shostname%[1]s is where you log in to GitHub. The default hostname is %[1]sgithub.com%[1]s.
```
"log in" might be a touch confusing but I don't feel that strongly.

jtmcg replied:
Do you have a better word/phrase in mind?

williammartin replied:
Not really. Maybe...

```
The default hostname is github.com, and can be overridden with the hostname flag
```

Maybe just leave it up to the reader to infer the URL?

jtmcg replied:
Let me ask a better question: why do you think this might be a touch confusing?

williammartin replied:
This is where I might log into GitHub https://github.com/login/

(Sorry short reply, hope it makes sense)

jtmcg replied:
No worries. It does make sense... 🤔

jtmcg replied:
Given this, I like your suggestion. Adopted 👍

williammartin commented on pkg/cmd/auth/login/login_test.go:
```diff
@@ -803,3 +804,50 @@ func Test_loginRun_Survey(t *testing.T) {
 	}
 }
+
+func Test_promptForHostname(t *testing.T) {
```
Interesting test!

williammartin reviewed (changes requested):
Deeply annoying that you can't comment on unmodified lines but I think we should adjust this block https://github.com/cli/cli/blob/ccb830ca3674dc480cae08f026f3e298662e35e1/pkg/cmd/auth/login/login.go#L236-L243

Firstly, `isEnterprise` doesn't capture the intent which is now "other". Secondly, there's no good reason to go lookup the hostname if we're not using it. Thirdly, it's idiomatic to early return in Go. Fourthly, `return thing, err` in the case where you know `err` is nil forces the developer to parse any previous conditionals in order to understand that there is no error.

I'd suggest changing this to:

```go
	isGitHubDotCom := hostType == 0
	if isGitHubDotCom {
		return ghinstance.Default(), nil
	}

	return opts.Prompter.InputHostname()
```

I'm not sure why we bother calling for the default hostname when we know the user chose github.com in the select but it also doesn't _really_ matter.

jtmcg committed:
Change conditional in promptForHostname to better reflect prompter changes

The old isEnterprise check no longer makes sense, given the prompter is
providing 'other', not 'GitHub Enterprise Server' as its non-GitHub.com
option. Additionally, there was an opportunity for cleaning up the code
via early returns and the removal of the default hostname lookup if we
don't need it.

jtmcg committed:
Update docs language to remove possible confusion around 'where you log in'

jtmcg commented:
> **Given** I am in an interactive terminal environment **When** I run gh auth login **Then** I am presented with:
> 
> ```
> ➜  gh auth login
> ? Where do you use GitHub? [Use arrows to move, type to filter]
>   github.com
> > other
> 
> ? Hostname:
> ```
> 
> ```
> ➜ ./bin/gh auth login
> ? Where do you use GitHub?  [Use arrows to move, type to filter]
>   GitHub.com
> > Other
> ```
> 
> ```
> ➜ ./bin/gh auth login
> ? Where do you use GitHub? Other
> ? Hostname:
> ```

Is there a change requested by this comment, @williammartin?

williammartin commented:
> Is there a change requested by this comment, @williammartin?

No, sorry, it was me doing the A/C and then was following up with the review but I didn't comment it as such.

williammartin reviewed (approved):
LGTM!
